doc/dotNET SDK Proposal.md

```diff
@@ -0,0 +1,232 @@
+
+The `MainAsync`-method will run and eventually invoke the `ExecuteAsync(..)` method to request the execution of `SomeActivity`. As [discussed earlier](#DotNetAsyncIsLogicallySingleThreaded), this invocation occurs completely synchronously. `ExecuteAsync` will make a note that a command to schedule `SomeActivity` must be issued when the current workflow Task eventually completes, and it will return a Task that will complete when the activity scheduled in that way completes in some future.
+
+After `ExecuteAsync` returns, the workflow chooses to await the returned Task, i.e. the completion of `SomeActivity`. When the Task is awaited, the normal .NET async/await mechanism will check whether the awaited Task is completed. (Awaiting completed Tasks completes synchronously and immediately in .NET.) However, the workflow task is still running, and the Temporal server does not yet even know that the activity has been scheduled. So, the activity Task is not completed. Thus, the .NET async/await mechanism will capture the local state and return the execution flow to the caller, i.e. to the SDK Engine message loop. 
+
+Any time the execution returns to the Engine's message loop in this way, the ongoing workflow task is completed. The Engine will call Core to notify the Temporal server of the workflow task completion and it will include all commands issued during the task execution. In this case, it will include the command to schedule `SomeActivity`. After that, the message loop will continue.
+
+The Temporal server will perform the necessary orchestration to perform `SomeActivity`. When it is eventually completed, the respective information will appear in the workflow history, and it will be picked up from `SomeTaskQueue` by Core and then by the Engine message loop.
+
+The Engine will use its internal local state to look up the Task object that represents the completion of `SomeActivity`. Recall that it is the object previously returned from `ExecuteAsync`. Completing that Task will cause the normal .NET async/await mechanism to resume the execution of `MainAsync` where it left off previously.
+
+During the next loop iteration, the same procedure will occur. Eventually, the required condition will be met and some client will invoke the `NotifyConditionMet` signal. Once this information appears in the workflow history, Core will pick it up and make a corresponding activation available for the Engine's message loop. The engine will pick up the activation and invoke the `NotifyConditionMetAsync()`-method. In turn, it will set `_isConditionMet` to `true`. During the subsequent `MainAsync` iteration the loop will exit and the workflow will complete.
+
+### Workflow parameters
+
+A .NET workflow can only receive one parameter. The parameter type must implement the `IDataValue` interface. That interface does not specify any members. Note that any number of parameters can be encoded as fields on a type that implements `IDataValue`.
```

Comment:
I think requiring parameters and return types to implement some interface is very reasonable. In Rust, it'll likely be the `sede` `Serialize` and `Deserialize` types - however, those interfaces actually require you to implement that functionality and they also are automatically implemented for 90% of types you'd ever use.

I'm not sure I follow what the value is of requiring implementing an *empty* interface, and I definitely agree with Chad that if such an interface cannot automatically be implemented for primitives / stdlib types then it starts to have negative value.

Reply:
> such an interface cannot automatically be implemented

That is the case here same as in Java. I think, like all of our other SDKs, you should be able to fallback to default language-supported JSON serialization of the types in either direction.

Reply:
Yes, that's exactly what I mean. The interface should be automatically implemented for any type which can use such built-in serialization. In Rust it's as easy as slapping a `#[derive(Serialize, Deserialize)]` attribute to your type, and if it can't be serialized then it fails to compile. That's pretty ideal.

But, requiring that annotation is nice because now I can break at compile time rather than runtime if you try to use an non-serializable struct as an arg/return type. If this cannot be done at compile time here, then there isn't much value. Just attempt to serialize at runtime and blow up if it doesn't work.

Reply:
Like all of the existing SDKs, the language supports serializing regular objects to JSON and C# shouldn't be different. Rust is a bit different w/ its derivation and it's general-purpose serde usage. In these other languages, it's the norm to use regular objects, potentially utilizing metadata (Java annotations, Go tags, C# attributes, etc) to customize JSON serialization as needed.

Reply:
The purpose of the empty iface `IDataValue` is that the json-serialize parameter is an object. Once the parameter is an object, it becomes trivial to add fields to in in a backward-compatible manner. On contrary, if you originally just passed an int or a string, you cannot simply add another field.

Reply:
I'll add this clarification to the doc